cachi2/core/package_managers/gomod.py

```diff
@@ -660,228 +670,272 @@ def _should_vendor_deps(
+        self,
+        module_name: str,
+        app_dir: RootedPath,
+    ) -> str:
+        """
+        Get the version of the Go module in the input Git repository in the same format as `go list`.
 
-    log.debug("No semantic version tag was found on the commit %s", commit_sha)
+        If commit doesn't point to a commit with a semantically versioned tag, a pseudo-version
+        will be returned.
 
-    # This logic is based on:
-    # https://github.com/golang/go/blob/a23f9afd9899160b525dbc10d01045d9a3f072a0/src/cmd/go/internal/modfetch/coderepo.go#L511-L521
-    for major_version in major_versions_to_try:
-        # Get the highest semantic version tag before the commit with a matching major version
-        pseudo_base_tag = _get_highest_semver_tag(
-            repo, commit, major_version, all_reachable=True, subpath=subpath
-        )
-        if not pseudo_base_tag:
-            continue
+        :param module_name: the Go module's name
+        :param app_dir: the path to the module directory
+        :return: a version as `go list` would provide
+        """
+        # If the module is version v2 or higher, the major version of the module is included as /vN at
+        # the end of the module path. If the module is version v0 or v1, the major version is omitted
+        # from the module path.
+        module_major_version = None
+        match = re.match(r"(?:.+/v)(?P<major_version>\d+)$", module_name)
+        if match:
+            module_major_version = int(match.groupdict()["major_version"])
+
+        if module_major_version:
+            major_versions_to_try: tuple[int, ...] = (module_major_version,)
+        else:
+            # Prefer v1.x.x tags but fallback to v0.x.x tags if both are present
+            major_versions_to_try = (1, 0)
 
-        log.debug(
-            "Using the semantic version tag of %s as the pseudo-base for the commit %s",
-            pseudo_base_tag.name,
-            commit_sha,
-        )
-        pseudo_version = _get_golang_pseudo_version(
-            commit, pseudo_base_tag, major_version, subpath=subpath
+        if app_dir.path == app_dir.root:
+            subpath = None
+        else:
+            subpath = app_dir.path.relative_to(app_dir.root).as_posix()
+
+        for major_version in major_versions_to_try:
+            # Get the highest semantic version tag on the commit with a matching major version
+            tag_on_commit = self._get_highest_semver_tag(major_version, subpath=subpath)
+            if not tag_on_commit:
+                continue
+
+            log.debug(
+                "Using the semantic version tag of %s for commit %s",
+                tag_on_commit.name,
+                self._commit.hexsha,
+            )
+
+            # We want to preserve the version in the "v0.0.0" format, so the subpath is not needed
+            return (
+                tag_on_commit.name if not subpath else tag_on_commit.name.replace(f"{subpath}/", "")
+            )
```

Comment:
Since we're also doing a refactoring in this PR, another improvement suggestion is to break this piece here into a separate function (`_get_highest_semver_tag_on_current_commit`?), as well as the other piece that is currently L798-817 (`_get_highest_reachable_semver_tag`?). 

I always find it hard to read this function here, maybe breaking it a little would make it a little easier on the eyes. But again, just a suggestion, ignore it if you don't like it.

Reply:
I'll take a look at this today :+1:

Reply:
Good call, I think it's more readable now